content/blog/2024-01-11-xilem-backend-roadmap.md

```diff
@@ -0,0 +1,157 @@
+
+And here's the thing: realistically, most layout engines aren't that creative. Every Rust framework and every major GUI framework I'm aware of uses layout that boils down to "FlexBox with small variations". The number of meaningfully different ways to do GUI layout can be counted on, generously, two hands.
+
+I'd argue we only need CSS Flow, FlexBox, CSS Grid, and maybe Morphorm to cover most people's needs. Conveniently, this is what Taffy aims to provide. There's already an experimental Taffy integration in Xilem, and I want to make it deeper. More on this in a latter blog post.
+
+Also, having a fixed set of widgets mean you can avoid using `Box<dyn Widget>`, which has some interesting performance implications. It means you can add a lot of methods to the Widget trait without worrying about bloating the vtable. And you can write methods under the assumption that they'll be inlined, which opens up interesting DX improvements.
+
+
+### Using ECS, maybe, sort of
+
+In current Xilem code, if you want your FlexWidget to have children, it must have a `Vec<Pod<ChildWidget>>` member. Each Pod directly owns a Widget.
+
+I think this is an interesting quirk of Rust GUI, tying directly to Rust's ownership semantics. Non-Rust frameworks often use pointers to child widgets, true, but they're often shared with the framework, with an understanding that the framework is charged with keeping track of how many widgets there are and which is a child of which. In Qt, for instance, it's preferred (but not mandatory) for parents to call `delete_later()` on their children instead of `delete`, because it plays better with event handling.
+
+Xilem's approach of "you own your children" is a little bespoke. It means the framework has a hard time iterating over the entire widget tree (say, as part of a DOM inspector) unless each Widget specifically implements the iterating method. And it means if you want to target an event at a specific widget (say, the one with keyboard focus), you need to go through the entire ownership chain, using some heuristics to avoid visiting too many widgets.
+
+I believe Widgets should be owned by the library. If your container has children, then the only thing the container will actually own is keys into a structure (probably a slotmap) where the widget is stored. This makes a lot of things easier, like serialization and debugging, but it has an impact on the entire backend. It's an infrastructure investment.
+
+Lately, I've seen more and more dicussion of implementing GUI through an ECS. A lot of that discussion comes from Bevy, which is natural, since the bevy community ~~is filled with ruthless cultists striving to feed ever more sacrifices to the ECS god~~
```

Comment:
My experience with these kinds of jokes is that some fraction of the readers will not get the context and will take it literally.

Reply:
I've been told by Alice that the Bevy community found it funny =D

Reply:
(Like, it was posted as a screenshot on the \#memes channel, which I find a bit unsettling, but we cool.)